src/main/java/com/amazon/opendistroforelasticsearch/sql/request/SqlRequestFactory.java

```diff
@@ -59,16 +61,36 @@ private static SqlRequest parseSqlRequestFromPayload(RestRequest restRequest) {
         JSONObject jsonContent;
         try {
             jsonContent = new JSONObject(content);
+            if (jsonContent.has(SQL_CURSOR_FIELD_NAME)) {
+                return new SqlRequest(jsonContent.getString(SQL_CURSOR_FIELD_NAME));
+            }
         } catch (JSONException e) {
             throw new IllegalArgumentException("Failed to parse request payload", e);
         }
         String sql = jsonContent.getString(SQL_FIELD_NAME);
+
         if (jsonContent.has(PARAM_FIELD_NAME)) { // is a PreparedStatement
             JSONArray paramArray = jsonContent.getJSONArray(PARAM_FIELD_NAME);
             List<PreparedStatementRequest.PreparedStatementParameter> parameters = parseParameters(paramArray);
             return new PreparedStatementRequest(sql, jsonContent, parameters);
         }
-        return new SqlRequest(sql, jsonContent);
+        return new SqlRequest(sql, validateAndGetFetchSize(jsonContent), jsonContent);
+    }
+
+
+    private static Integer validateAndGetFetchSize(JSONObject jsonContent) {
+        Integer fetchSize = null;
+        try {
+            if (jsonContent.has(SQL_FETCH_FIELD_NAME)) {
+                fetchSize = jsonContent.getInt(SQL_FETCH_FIELD_NAME);
+                if (fetchSize < 0) {
+                    throw new IllegalArgumentException("Fetch_size must be greater or equal to 0");
+                }
+            }
+        } catch (JSONException e) {
+            throw new IllegalArgumentException("Failed to parse field [" + SQL_FETCH_FIELD_NAME +"]", e);
```

Comment:
In which case the exception will been throw? Does it been tested?

Reply:
This exception will be thrown when passing non-numeric value to `fetch_size` , for example `"fetch_size" : "hello world"`. An `IllegalArgumentException` is thrown to be caught as client error (400) in `RestSqlAction.java`. Will add a test for this.